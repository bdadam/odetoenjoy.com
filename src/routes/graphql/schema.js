import { buildSchema } from 'graphql';

export default buildSchema(`
schema {
    query: Query,
    mutation: Mutation
}

type Query {
    healthcheck: String

    crawlUrl(url: String!): CrawlResult
}

type Video2 {
    id: ID!
    title: String


    # resolution: Resolution
    # url: String
    # embed: String
}

type MediaUrls {
    canonical: String
    embed: String
}

type Resolution {
    width: Int
    height: Int
}

type Mutation {
    createVideo(videoInput: VideoInput!): Video

    suggestVideo(input: VideoSuggestionInput!): VideoSuggestionResult!
}

input VideoSuggestionInput {
    url: String!
    title: String
    description: String
    tags: [String]    
}

type VideoSuggestionResult {
    accepted: Boolean!
    url: String
}

type VideoPage {
    id: ID!
    slug: String!
    title: String!
    description: String
    tags: [String]
    videos: [Video]
}

type Video {
    previewImage: Image
    embedUrl: String!
    duration: Int
}

type Image {
    id: ID
}

type PageMetaData {
    title: String!
    description: String
    canonicalUrl: String!
}

input VideoInput {
    url: String!
    title: String!
    description: String
    tags: [String]
}

type VideoPageSummary {
    id: String!
    url: String!
    title: String!
}

type CrawlResult {
    url: String!
    canonicalUrl: String
    title: String
    videos: [CrawledVideo]
    images: [CrawledImage]
}

type CrawledVideo {
    url: String,
    width: Int,
    height: Int,
    type: String
}

type CrawledImage {
    url: String,
    width: Int,
    height: Int
}
`);
