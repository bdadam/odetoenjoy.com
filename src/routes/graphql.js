import { Router } from 'express';

const graphqlHTTP = require('express-graphql');
const { buildSchema, graphql } = require('graphql');

const router = Router();

const schema = buildSchema(`
    schema {
        query: Query,
        mutation: Mutation
    }

    type Query {
        crawlUrl(url: String!): CrawlResult
        # findPage
    }

    type Mutation {
        createVideo(videoInput: VideoInput!): Video

        suggestVideo(input: VideoSuggestion!): VideoSuggestionResult
        
    }

    input VideoSuggestion {
        url: String!
        title: String!
        description: String
        tags: [String]
        
    }

    type VideoSuggestionResult {
        accepted: Boolean!
        url: String
    }

    type VideoPage {
        id: ID!
        slug: String!
        title: String!
        description: String
        tags: [String]
        videos: [Video]
    }

    type Video {
        previewImage: Image
        embedUrl: String!
        duration: Int
    }

    type Image {
        id: ID
    }

    type PageMetaData {
        title: String!
        description: String
        canonicalUrl: String!
    }

    input VideoInput {
        url: String!
        title: String!
        description: String
        tags: [String]
    }

    type VideoPageSummary {
        id: String!
        url: String!
        title: String!
    }

    type CrawlResult {
        url: String!
        canonicalUrl: String
        title: String
        videos: [CrawledVideo]
        images: [CrawledImage]
    }

    type CrawledVideo {
        url: String,
        width: Int,
        height: Int,
        type: String
    }

    type CrawledImage {
        url: String,
        width: Int,
        height: Int
    }
`);

// import { createPreviewFromUrl } from '../services/image-service';

// createPreviewFromUrl('https://i.ytimg.com/vi/fW8amMCVAJQ/hqdefault.jpg').then(x => console.log(x));

// import Video from '../models/video-model';

// Video.create({
//     title: 'TEST',
//     description: 'sdfsfsdf'
// }).then(v => {
//     console.log(v);
// });

// Video.all().then(vids => {
//     console.log(vids);
// });

// Promise.race([fetch('...'), new Promise(resolve => setTimeout(resolve, 2000))]).then(...);

import VideoPostService from '../services/video-post-service';

const root = {
    crawlUrl: (args, ctx) => {
        return VideoPostService.fetchMetaData(args.url);
    },
    addVideo: (args, ctx) => {
        console.log(args);

        return new Promise((resolve, reject) => {
            resolve({
                id: args.id || '123',
                url: args.videoInput.url || 'sdfdsf',
                title: args.videoInput.title || 'sfdsdfsd'
            });
        });
    }
};

// const ctx = { test: 123 };
// const r = graphql(schema, '{ text }', root, ctx);
// r.then(x => console.log(x));

router.use('/', graphqlHTTP({
    schema: schema,
    rootValue: root,
    graphiql: true,
}));

export default router;
