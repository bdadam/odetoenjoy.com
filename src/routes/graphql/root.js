import VideoPostService from '../../services/video-post-service';

const root = {
    healthcheck: () => {
        return 'OK';
    },
    crawlUrl: (args, ctx) => {
        return VideoPostService.fetchMetaData(args.url);
    },
    addVideo: (args, ctx) => {
        console.log(args);

        return new Promise((resolve, reject) => {
            resolve({
                id: args.id || '123',
                url: args.videoInput.url || 'sdfdsf',
                title: args.videoInput.title || 'sfdsdfsd'
            });
        });
    },
    suggestVideo: (args) => {
        // console.log(args);
        return Promise.resolve({ accepted: true, url: 'blah' });
    }
};

// import { createPreviewFromUrl } from '../services/image-service';

// createPreviewFromUrl('https://i.ytimg.com/vi/fW8amMCVAJQ/hqdefault.jpg').then(x => console.log(x));

// import Video from '../models/video-model';

// Video.create({
//     title: 'TEST',
//     description: 'sdfsfsdf'
// }).then(v => {
//     console.log(v);
// });

// Video.all().then(vids => {
//     console.log(vids);
// });

// Promise.race([fetch('...'), new Promise(resolve => setTimeout(resolve, 2000))]).then(...);

export default root;
