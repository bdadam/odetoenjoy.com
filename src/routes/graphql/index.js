import { Router } from 'express';
import graphqlHTTP from 'express-graphql';

import schema from './schema';
import root from './root';

const router = Router();

// const ctx = { test: 123 };
// const r = graphql(schema, '{ text }', root, ctx);
// r.then(x => console.log(x));

router.use('/', graphqlHTTP({
    schema: schema,
    rootValue: root,
    graphiql: true,
    formatError: error => ({
        message: error.message,
        locations: error.locations,
        stack: error.stack,
        path: error.path
    })
}));

export default router;
