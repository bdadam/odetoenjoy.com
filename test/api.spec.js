import { graphql } from 'graphql';
import schema from '../src/routes/graphql/schema';
import root from '../src/routes/graphql/root';

import request from 'supertest';
import app from '../src/app';



const queryApp = (query, variables, method = 'get') => {
    return request(app)[method](`/gql`).query({ query, variables: JSON.stringify(variables) }).set('Accept', 'application/json');
};

describe('Video Suggestions', () => {
    it('Happy case', () => {
        const mutation = `
            mutation VideoSuggestion($input: VideoSuggestionInput!) {
                suggestVideo(input: $input) {
                    accepted
                    url
                }
            }
        `;

        const input = {
            url: 'some-very-good-video-url',
            title: 'A very good title',
            description: 'Some description',
            tags: ['tag1', 'tag2', 'tag3']
        };

        // const x = 'xxxxxxxx'.replace(/x/g, () => {
        //     const alphabet = 'abcdef';
        //     return alphabet[Math.random() * alphabet.length | 0];
        // });
        // console.log(x);

        return queryApp(mutation, { input }, 'post').then(response => {
            expect(response.statusCode).toEqual(200);
            expect(response.body).toMatchSnapshot();
        });
    });
});
// jest.mock('../src/services/video-post-service.js', () => {
//     return {
//         fetchMetaData(url) {
//             switch(url) {
//                 case 'fine-url':
//                     return Promise.resolve({
//                         title: 'TITLE',
//                         canonicalUrl: `canonical-${url}`
//                     });
//             }
//         }
//     };
// });

describe('Test', () => {
    it('FetchMetaData', () => {

        const query = `
            query ($url: String!) {
                crawlUrl(url: $url) {
                    title
                    canonicalUrl
                    videos {
                        url
                        type
                    }
                }
            }
        `;
    
        return queryApp(query, { url: 'fine-url' }).then(response => {
            expect(response.statusCode).toBe(200);
            expect(response.body).toMatchSnapshot();
        });
    });
});

// jest.mock('../src/databases/videodb');

import videoModel from '../src/models/video-model';
jest.mock('../src/databases/videodb');

describe('Post video', () => {
    beforeEach(() => {

    });

    it('bbb', () => {
        videoModel.create({
            title: 'title'
        });
    });    
});
