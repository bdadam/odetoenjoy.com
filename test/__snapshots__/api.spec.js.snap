// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test FetchMetaData 1`] = `
Object {
  "data": Object {
    "crawlUrl": Object {
      "canonicalUrl": "canonical-fine-url",
      "title": "Some og title",
      "videos": null,
    },
  },
}
`;

exports[`Video Suggestions Happy case 1`] = `
Object {
  "data": Object {
    "suggestVideo": Object {
      "accepted": true,
      "url": "blah",
    },
  },
}
`;
